internal/services/topic.go

```diff
@@ -312,14 +314,55 @@ func (ts *TopicService) requestedAssignments(currentPartitions int, brokers []*s
 		return brokers[i].ID() < brokers[j].ID()
 	})
 
-	assignments := make(map[int32][]int32, int(partitions))
+	rackMap := make(map[string][]*sarama.Broker)
+	var rackNames []string
+	brokersWithRack := 0
+	for _, broker := range brokers {
+		if broker.Rack() != "" {
+			brokersWithRack++
+			if _, ok := rackMap[broker.Rack()]; !ok {
+				rackMap[broker.Rack()] = make([]*sarama.Broker, 0)
+				rackNames = append(rackNames, broker.Rack())
+			}
+			rackMap[broker.Rack()] = append(rackMap[broker.Rack()], broker)
+		}
+	}
+
+	if len(brokers) != brokersWithRack {
+		if brokersWithRack > 0 {
+			glog.Warningf("Not *all* brokers have rack assignments (%d/%d), topic %s will be created without rack awareness", brokersWithRack, len(brokers), ts.canaryConfig.Topic)
+		}
+	} else {
+		index := 0
+
+		for ;; {
+			again := false
+
+			for _, rackName := range rackNames {
+				brokerList := rackMap[rackName]
+				if len(brokerList) > 0 {
+					var head *sarama.Broker
+					head, rackMap[rackName] = brokerList[0], brokerList[1:]
+					brokers[index] = head
+					index++
+					again = true
+				}
+			}
+
+			if !again {
+				break
+			}
```

Comment:
AFAIU this block is going to change the brokers order so that we have one after the other brokers in different racks.
For example, let's assume this scenario:

rack "es1" --> brokers 0,1
rack "es2" --> brokers 3,5
rack "es3" --> brokers 2,4

the final brokers list will be something like this:

```shell
es1 | es2 | es3 | es1 | es2 | es3
-----------------------------------------
  0   |   3   |   2  |   1   |   5   |   4
-----------------------------------------
```

Is my understanding right? Can we add a comment (ASCII art example?) to make it clear?

Reply:
> AFAIU this block is going to change the brokers order so that we have one after the other brokers in different racks.
That's right.

> Is my understanding right? Can we add a comment (ASCII art example?) to make it clear?

Good idea, I'll do it.